app/code/Magento/Catalog/Block/Product/ListProduct.php

```diff
@@ -466,6 +466,26 @@ private function initializeProductCollection()
 
         $this->addToolbarBlock($collection);
 
+        $restrictedIds = [];
+        $prodCollcetion = clone $collection;
```

Comment:
Why do you need to keep in memory the whole copy of products collection? Such changes may cause huge performance degradation